### PR TITLE
semantic highlighting: add reference hlmod

### PR DESCRIPTION
**Pull request by jhgg**
This PR adds the "reference" highlight modifier! 

I basically went around and looked for `HlMod::Mutable` to find the callsites to add a reference. I think these all make sense!

### CODE REVIEW

**Comment by matklad** [comment]
I agree that something along this lines makes sense, but I am not sure that we want quite exactly this. 

THe current impleentation treats mutable/refernce as a comulimentary way, but mutability and reference-ness are orthogonal concepts. Here are all 4 cases:

```rust
let         x
let     mut x
let ref     x
let ref mut x
```

Question: what exactly do you want to see in terms of actual syntax hightlighting?  Do you want to see indirection? Or do you want to see non-mutability?

**Comment by jhgg** [comment]
I want to make variables that are things I don't own show differently. So, I think "indirection"

File: crates/ide/src/syntax_highlighting/highlight.rs [hunk 1]
```diff
@@ -487,6 +489,8 @@ fn highlight_def(db: &RootDatabase, krate: Option<hir::Crate>, def: Definition)
             let ty = local.ty(db);
             if local.is_mut(db) || ty.is_mutable_reference() {
                 h |= HlMod::Mutable;
+            } else if local.is_ref(db) || ty.is_reference() {
+                h |= HlMod::Reference;
             }
```
Comment (matklad) [inline crates/ide/src/syntax_highlighting/highlight.rs:494]:
```suggestion
            } 
            if local.is_ref(db) || ty.is_reference() {
                h |= HlMod::Reference;
            }
```

let's do this here and elswhere, to keep mutability and reference-ness orthogonal then!

File: crates/hir/src/lib.rs [hunk 2]
```diff
@@ -2195,6 +2200,10 @@ impl Type {
         matches!(self.ty.kind(&Interner), TyKind::Ref(hir_ty::Mutability::Mut, ..))
     }
 
+    pub fn is_reference(&self) -> bool {
+        matches!(self.ty.kind(&Interner), TyKind::Ref(hir_ty::Mutability::Not, ..))
```
Comment (Veykril) [inline crates/hir/src/lib.rs:2204]:
```suggestion
        matches!(self.ty.kind(&Interner), TyKind::Ref(..))
```
This should ignore the mutability

File: crates/hir/src/lib.rs [hunk 3]
```diff
@@ -1840,6 +1840,11 @@ impl Local {
         matches!(&body[self.pat_id], Pat::Bind { mode: BindingAnnotation::Mutable, .. })
     }
 
+    pub fn is_ref(self, db: &dyn HirDatabase) -> bool {
+        let body = db.body(self.parent);
+        matches!(&body[self.pat_id], Pat::Bind { mode: BindingAnnotation::Ref, .. })
```
Comment (Veykril) [inline crates/hir/src/lib.rs:1845]:
```suggestion
        matches!(&body[self.pat_id], Pat::Bind { mode: BindingAnnotation::Ref | BindingAnnotation::RefMut, .. })
```

File: crates/ide/src/syntax_highlighting/highlight.rs [hunk 4]
```diff
@@ -375,11 +375,11 @@ fn highlight_def(db: &RootDatabase, krate: Option<hir::Crate>, def: Definition)
                 if let Some(item) = func.as_assoc_item(db) {
                     h |= HlMod::Associated;
                     match func.self_param(db) {
-                        Some(sp) => {
-                            if let hir::Access::Exclusive = sp.access(db) {
-                                h |= HlMod::Mutable;
-                            }
-                        }
+                        Some(sp) => match sp.access(db) {
+                            hir::Access::Exclusive => h |= HlMod::Mutable,
```
Comment (Veykril) [inline crates/ide/src/syntax_highlighting/highlight.rs:379]:
```suggestion
                            hir::Access::Exclusive => h = h | HlMod::Mutable | HlMod::Reference,
```
(can't use `h |= ...` due to a missing impl for that I believe)

File: crates/ide/src/syntax_highlighting/highlight.rs [hunk 5]
```diff
@@ -460,6 +460,8 @@ fn highlight_def(db: &RootDatabase, krate: Option<hir::Crate>, def: Definition)
                 if s.is_mut(db) {
                     h |= HlMod::Mutable;
                     h |= HlMod::Unsafe;
+                } else {
+                    h |= HlMod::Reference;
```
Comment (Veykril) [inline crates/ide/src/syntax_highlighting/highlight.rs:464]: non mutable statis aren't references so this seems wrong to me.

File: crates/ide/src/syntax_highlighting/highlight.rs [hunk 6]
```diff
@@ -549,7 +554,7 @@ fn highlight_method_call(
 
     if let Some(self_param) = func.self_param(sema.db) {
         match self_param.access(sema.db) {
-            hir::Access::Shared => (),
+            hir::Access::Shared => h |= HlMod::Reference,
             hir::Access::Exclusive => h |= HlMod::Mutable,
```
Comment (Veykril) [inline crates/ide/src/syntax_highlighting/highlight.rs:558]:
```suggestion
            hir::Access::Exclusive => h | HlMod::Mutable | HlMod::Reference,
```

**Comment by jhgg** [comment]
Thanks for the suggests. Applied all of them & updated tests!

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**